[TSan] Add dominance-based redundant instrumentation elimination

apaznikov opened the pull request:
## Summary

ThreadSanitizer instruments every memory access with a runtime call. If
access A dominates access B, both access the same memory location, and no
synchronization exists on any CFG path between them, B's instrumentation is
redundant — any race that B would detect is already detected at A. This PR
implements a new pass that identifies and removes such redundant checks.

This work is part of a broader research effort on optimizing dynamic race
detectors [1].

## Motivation

Our experiments identified dominance-based elimination as the single most
effective optimization for reducing TSan overhead. In our research prototype
(which uses a more aggressive, inter-procedural version of this analysis),
we observed the following speedups from this optimization alone:

* **Redis:** 1.35x speedup
* **FFmpeg:** 1.30x speedup
* **SQLite:** 1.21x speedup
* **MySQL Select:** 1.15x speedup
* **MySQL Write-only:** 1.11x speedup
* **Memcached:** 1.03x speedup

**Chromium** — DE-specific results on representative benchmarks:
1.06x on Speedometer 3.1, 1.19x on Parser: HTML5 Render, 1.20x on
SVG: SvgCubics, 1.35x on Paint: Transform Changes, 1.34x on Image
Decoder: WebP.

This patch implements a **conservative, intra-procedural** version.
Absolute speedups may be lower than the prototype, but the pass targets
the same redundancy patterns and provides the infrastructure for future
improvements.

## Correctness

Access B is eliminated when all of the following hold:

1. A dominates B (A executes on every path reaching B).
2. A and B MustAlias, and A covers at least as many bytes as B.
3. If B is a write, A must also be a write (a dominating read cannot
   subsume a subsequent write).
4. No synchronization (atomic, fence, or call without `nosync`) exists on
   any CFG path from A to B.

When all four conditions hold, any race detectable at B is already
detectable at A. The full argument and implementation details are in the
`DominanceBasedElimination` class comments in `ThreadSanitizer.cpp`.

## Report Granularity

In some cases the reported race location shifts from the dominated access
to the dominating one. The presence of a race is always reported.

## This PR

This is the first in a series:

1. **This PR:** dominance direction only, enabled by default.
2. **Follow-up:** post-dominance extension — if B post-dominates A and
   the path is safe, A's check is redundant.

## Attribution

Research and algorithm design:
Alexey Paznikov, Michael Schwarz, Andrey Kogutenko, Yaroslav Osipov, Umang Mathur.

[1] Preprint: https://arxiv.org/abs/2512.05555

apaznikov committed:
[TSan] Add dominance-based redundant instrumentation elimination

This patch introduces a new optimization pass to ThreadSanitizer that eliminates redundant instrumentation of memory accesses using dominance and post-dominance analysis.

The optimization relies on the observation that if two memory accesses to the same location are executed on a path without intermediate synchronization (e.g., atomic operations, fences, or unknown function calls), checking one of them is often sufficient to detect a data race.

The algorithm works in two phases:
1. Dominance-based elimination: If access A dominates access B, they target the same memory location (MustAlias), and the path from A to B is free of synchronization, the instrumentation for B is redundant.
2. Post-dominance-based elimination: If access B post-dominates access A, they target the same location, and the path from A to B is safe, the instrumentation for A can be removed (since execution is guaranteed to reach B, where the race will be detected).

To ensure soundness and prevent false negatives:
- We use strict MustAlias results from AliasAnalysis.
- We perform a safety analysis to ensure no "dangerous" instructions (synchronization, unknown calls) exist between the accesses.
- For post-dominance, we conservatively avoid elimination across loops to prevent issues with non-terminating loops where the post-dominating access might never be reached.

This implementation is intra-procedural and utilizes standard LLVM analyses (DominatorTree, PostDominatorTree, AAResults, LoopInfo).

This optimization is disabled by default and can be enabled via `-tsan-use-dominance-analysis`.

This implementation is based on the research algorithm designed by:
Alexey Paznikov, Andrey Kogutenko, Yaroslav Osipov, Michael Schwarz, and Umang Mathur.

dvyukov commented on llvm/lib/Transforms/Instrumentation/ThreadSanitizer.cpp:
```diff
@@ -84,6 +88,15 @@ static cl::opt<bool>
     ClOmitNonCaptured("tsan-omit-by-pointer-capturing", cl::init(true),
                       cl::desc("Omit accesses due to pointer capturing"),
                       cl::Hidden);
+static cl::opt<bool>
```
It would be useful to have a meaningful description of what exactly this is doing somewhere in the source code (probably here). The description may be condensed from the PR description.
Just repeating the variable/flag name (dominance analysis) won't tell me much in future.
It also needs to say why this is disabled by default, and what's the plan for enabling it in future (are there any major unresolved problems, or it's just not enabled to do more gradual enabling). Note that if it's disabled by default, it won't really get lots of testing out there, and won't bring much benefit (almost nobody will enable this).
Also it needs to detail what's exactly unsafe in tsan-postdom-aggressive.

apaznikov committed:
[TSan] Enable Dominance elimination by default and address review feedback

Prepare the Dominance-based elimination pass for upstreaming:
* Enable the optimization by default (-tsan-use-dominance-analysis=true).
* Add a detailed comment block explaining the dominance and post-dominance
  logic, as well as the strict safety guarantees.
* Remove the experimental aggressive post-dominance logic entirely to keep
  the implementation strictly sound and simple.
* Clean up custom lit test configurations.
* Explicitly disable the optimization in `capture-no-omit.ll` and
  `read_before_write.ll` to preserve existing test coverage.
* Rewrite `dominance-elimination.ll`: drop the `test_` prefix from
  function names, add NODOM and VOLATILE check prefixes, reorganize into
  labelled sections, and add new cases for post-dominance read/write
  interactions, multi-path dirty/clean, loop safety, call-attribute
  safety, synchronization barriers, and alias analysis.
* Fix the algorithm implementation

apaznikov commented:
Gentle ping. @dvyukov , @melver — I've updated this PR based on your suggestions:
1. The optimization is now enabled by default.
2. Added detailed documentation in the source code explaining the logic and safety guarantees.
3. Removed the experimental aggressive post-dom mode to ensure strict soundness.

apaznikov commented:
@fmayer @thurstond — Hi! Since Dmitry and Vitaly are likely swamped right now, would you mind taking a look at this TSan optimization? It eliminates redundant instrumentation based on CFG dominance. It's enabled by default, strictly sound, and reduces TSan overhead significantly. All tests are green. Thanks!

vitalybuka commented on llvm/lib/Transforms/Instrumentation/ThreadSanitizer.cpp:
```diff
@@ -670,16 +1254,27 @@ bool ThreadSanitizer::instrumentLoadOrStore(const InstructionInfo &II,
+    break;
+  // Not specified yet:
+  // case AtomicOrdering::Consume:                v = 1; break;
+  case AtomicOrdering::Acquire:
```
Please extract all unrelated refactorings, reformatting into NFC PRs
https://llvm.org/docs/CodeReview.html#code-reviews-speed-and-reciprocity

If you have ideas how to decompose main change in set of smaller obvious patches, it will be awesome.

apaznikov replied:
@vitalybuka Thank you for the review!
To address your feedback about decomposing the main change into smaller, obvious patches, I have performed a massive refactoring:

1. Removed Post-Dominance completely. This PR now strictly focuses on simple Dominance.
2. Removed path caching. The implementation is now much leaner and easier to verify.
3. Reverted all unrelated formatting changes (NFCs).
4. Moved the implementation details from the PR description directly into Doxygen comments above the class.

Note: Because the scope of the patch changed so drastically, I squashed the changes into a single clean commit and force-pushed. It should be much easier to review from scratch now.

Let me know if this simplified version looks better!

vitalybuka commented:
"Implementation Details" should be comments in the code, explaining relevant pieces.

PR description should focus on motivation, which looks good, but some parts are quite redundant.
E..g "Impact" vs "Motivation & Potential Impact"

fmayer commented:
> Usage
> The optimization is currently opt-in.
> Flag: -mllvm -tsan-use-dominance-analysis

This part of the commit message is incorrect.

fmayer commented:
IMO this is redundant vs what Git tracks anyway

> Implementation:
> This patch was implemented by Alexey Paznikov.

apaznikov committed:
[TSan] Add dominance-based redundant instrumentation elimination

Adds DominanceBasedElimination, a new optimization that removes redundant
TSan instrumentation when one memory access dominates another to the same
memory location with no synchronization on any CFG path between them.

If access A dominates access B, both target the same memory location
(MustAlias, with A covering at least as many bytes as B), and no
synchronization (atomic, fence, or call without nosync) exists on any
CFG path from A to B:
- A write covers both read and write races; B is redundant regardless of
  whether B is a read or a write.
- A read covers only read races; B is redundant only if B is also a read.

In our research prototype (which uses a more aggressive, inter-procedural
version of this analysis), we observed the following speedups from
dominance-based elimination alone:
* Redis: 1.35x, FFmpeg: 1.30x, SQLite: 1.21x
* MySQL Select: 1.15x, MySQL Write-only: 1.11x, Memcached: 1.03x
* Selected Chromium benchmarks: Speedometer 3.1 1.06x, Parser: HTML5
  Render 1.19x, SVG: SvgCubics 1.20x, Paint: Transform Changes 1.35x,
  Image Decoder: WebP 1.34x.

This patch implements a conservative, intra-procedural version; absolute
speedups may be lower than the prototype. Post-dominance elimination will
follow in a subsequent patch.

The optimization is enabled by default. Caching is intentionally omitted
to keep the implementation reviewable; it can be added as a follow-up if
compile-time profiling justifies it.

This work is part of a broader research effort on optimizing dynamic
race detectors. Preprint: https://arxiv.org/abs/2512.05555

Co-authored-by: Alexey Paznikov, Michael Schwarz, Andrey Kogutenko, Yaroslav Osipov, and Umang Mathur.

davidmrdavid commented:
Just swinging by from the LLVM discord
@dvyukov - it appears this PR could use another pass to unblock. Would you be able to take another look? Thanks!
